# componentDidCatch
- componentDidCatch 메서드
- 리액트 애플리케이션에서 발생하는 에러를 처리
- 함수형 컴포넌트에서는 구현할수 없는 기능이다.


#### 리액트 앱에서 에러날수있는 상황

`User.js`
- user props를 받아 사용자 정보를 출력하는 컴포넌트
```jsx
import React from 'react';

function User ({ user }) {
    return (
        <div>
            <div>
                <b>ID</b>: {user.id}
            </div>
            <div>
                <b>Username</b>: {user.username}
            </div>
        </div>
    );
};

export default User;
```

`App.js`
```jsx
return (
    <User />
);
```

App 컴포넌트에서 다음과 같이 User 컴포넌트를 사용할때, user props를 넘겨주지않으면 에러가 발생하게 된다.

사용자 입장에서는 에러인지 아닌지 알려주지도 않고, 흰 화면만 보이게 되는데 이렇게 되면 문제이다..

`컴포넌트 내에서 예외 처리를 통한 에러 처리`
```jsx
function User ({ user }) {
    // 유저가 없다면, 에러.
    // 하지만 아무것도 랜더링 하지 않음
    if (!user) return null;
    return (
        <div>
            <div>
                <b>ID</b>: {user.id}
            </div>
            <div>
                <b>Username</b>: {user.username}
            </div>
        </div>
    );
};
```

만약 user props가 존재하지 않는다면 null 을 리턴, 즉 아무것도 보여주지 않는다.

이 방법도 사용자에겐 에러라는 것을 알려주지 않기 때문에 문제가 있다.

만약 발견되지 않거나 예상치 못한 예외가 발생할 경우 이를 함수형 컴포넌트에서는 처리할 수가 없다.

클래스형 컴포넌트에서는 componentDidCatch 메서드를 활용하여 에러처리가 가능하다.

#### ErrorBoundary
```jsx
import React, { Component } from 'react';

// 에러를 처리할 ErrorBoundary 컴포넌트
class ErrorBoundary extends Component {
    state = {
        error: false,
    };

    componentDidCatch (error, info) {
        // error: 에러에 대한 정보
        // info: 에러가 난 곳에 대한 정보
        console.error('에러가 발생했습니다.');
        console.error(error, info);

        // 에러상태를 true로 변경
        this.setState({
            error: true,
        });
    }

    render () {
        // 에러 상태가 true면 에러 화면을 출력
        if (this.state.error) {
            return <h1>에러 발생 !!</h1>
        }
        return this.props.children;
    }
};

export default ErrorBoundary;
```

에러 처리를 위한 ErrorBoundary 컴포넌트 이다.

만약 에러가 발생한다면 componentDidCatch에서 에러 에 대한 정보와, 에러가 난 컴포넌트에 대한 정보를 알수 있다.

또한 state 를 활용하여 에러가 난경우 error페이지를 리턴 하도록 한다.

`ErrorBoundary를 활용한 에러 처리`
```jsx
return (
    <ErrorBoundary>
        <User />
    </ErrorBoundary>
);
```

ErrorBoundary 컴포넌트의 children 으로 User 컴포넌트를 넣어주면 User 컴포넌트에서 에러가 발생하더라도 ErrorBoundary로 인해 에러페이지가 보이게 된다.

